### PR TITLE
Some contributor documentation

### PR DESCRIPTION
**Pull request by parente**
More work on issue #517. Continuation of PR #538.

### CODE REVIEW

**Comment by jzf2101** [comment]
@parente I can review can you give me rights?

**Comment by parente** [comment]
@jzf2101 Thanks! Added the "Documentation" team with read/write access. 

There's one additional page I planned to include in this PR about community stacks managed outside this repository, so you'll see that .md file is blank at the moment.

File: docs/contributing/features.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# New Features
+
+We review contributions of new features (e.g., new packages, new scripts, new flags) to the core stacks in an attempt to balance the value of the images to the Jupyter community with the cost of maintaining the images over time. Roughly speaking, we think about the following when considering additions:
```
Comment (jzf2101) [inline docs/contributing/features.md]: IMHO, I'm not sure review is the right verb to use in this sentence because it seems as if the purpose of the review is "to balance the value of the images etc." rather than (what I think you mean) "we accept PR's to balance the value of the images etc."

Comment (jzf2101) [reply docs/contributing/features.md]: Also, I would consider stronger language than "think" since it's a means of evaluation

File: docs/contributing/features.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+* *Usefulness to Jupyter users* - Is the feature generally applicable across domains? Does it work with Jupyter Notebook, JupyterLab, JupyterHub, etc.?
```
Comment (jzf2101) [inline docs/contributing/features.md]: I think this would be easier to read with **bold**:  (bold followed by unbolded colon)

File: docs/contributing/features.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+* *Fit with the image purpose* - Does the feature match the theme of the stack in which it will be added? Would it fit better in a new, community stack?
+* *Complexity of build / runtime configuration* - How many lines of code does the feature require in one of the Dockerfiles or startup scripts? Does it require new scripts entirely? Do users need to adjust how they use the images?
+* *Impact on image metrics* - How many bytes does the feature and its dependencies add to the image(s)? How many minutes do they add to the build time?
+* *Ability to support the addition* - Can existing maintainers answer user questions and address future build issues? Are the contributors interested in helping with long-term maintenance? Can we write tests to ensure the feature continues to work over time?
+
+Follow the process below to suggest a new feature for inclusion in one of the core stacks:
```
Comment (jzf2101) [inline docs/contributing/features.md]: Please could we add "please"?

File: docs/contributing/features.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+1. Open a GitHub issue describing the feature you'd like to contribute.
```
Comment (jzf2101) [inline docs/contributing/features.md]: Probably unnecessary, but we could link here

File: docs/contributing/features.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+2. Discuss whether the addition makes sense in [one of the core stacks](../using/selecting.html#Core-Stacks), as a [recipe in the documentation](recipes.html), as a [community stack](stacks.html), or as something else entirely.
```
Comment (jzf2101) [inline docs/contributing/features.md]: Do we want to create a issue template as a followup?

File: docs/contributing/packages.md [hunk 6]
```diff
@@ -1,5 +1,14 @@
-# Packages
+# Package Updates
 
-## Package Updates
+We welcome all contributions that update packages already included in the project Dockerfiles. Follow the process below to update a package version:
```
Comment (jzf2101) [inline docs/contributing/packages.md]: Can we say please here?

File: docs/contributing/packages.md [hunk 7]
```diff
@@ -1,5 +1,14 @@
...
 
-## New Packages
\ No newline at end of file
+1. Locate the Dockerfile containing the library you wish to update.
```
Comment (jzf2101) [inline docs/contributing/packages.md]: Is there a way we can make this more specific as with the recipes file?

File: docs/using/recipes.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
+# Contributed Recipes
+
+Users sometimes share interesting ways of using the Jupyter Docker Stacks. We encourage users to [contribute these recipes](../contributing/recipes.html) to the documentation in case they prove useful to other members of the community. The sections below capture this knowledge.
+
+## Add RISE
+
+@pdonorio said:
```
Comment (jzf2101) [inline docs/using/recipes.md]: Is @ messaging going to properly cite when the docs are built in RST?

File: docs/using/recipes.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+> There is a great repo called [RISE](https://github.com/damianavila/RISE) which allow via extension to create live slideshows of your notebooks, with no conversion, adding javascript Reveal.js.
```
Comment (jzf2101) [inline docs/using/recipes.md]: FWIW I would attribute the users but reformat for simplicity

File: docs/using/recipes.md [hunk 10]
```diff
@@ -0,0 +1,256 @@
+# Contributed Recipes
```
Comment (jzf2101) [inline docs/using/recipes.md]: Is there a way we can highlight these in other places with links to these sections? I worry that while they're useful, people looking for a specific solution won't know to look here

File: docs/using/recipes.md [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+Sometimes it is useful to run the Jupyter instance behind a nginx proxy, for instance:
+
+- you would prefer to access the notebook at a server URL with a path (`https://example.com/jupyter`) rather than a port (`https://example.com:8888`)
+- you may have many different services in addition to Jupyter running on the same server, and want to nginx to help improve server performance in manage the connections
+
+Here is a [quick example NGINX configuration](https://gist.github.com/cboettig/8643341bd3c93b62b5c2) to get started.  You'll need a server, a `.crt` and `.key` file for your server, and `docker` & `docker-compose` installed.  Then just download the files at that gist and run `docker-compose up -d` to test it out.  Customize the `nginx.conf` file to set the desired paths and add other services.
+
+## Using spark-packages.org
+
+If you'd like to use packages from spark-packages.org, see [https://gist.github.com/parente/c95fdaba5a9a066efaab](https://gist.github.com/parente/c95fdaba5a9a066efaab) for an example of how to specify the package identifier in the environment before creating a SparkContext.
```
Comment (jzf2101) [inline docs/using/recipes.md]: Link for spark-packages?

File: docs/contributing/features.md (hunk 4 above)
Comment (parente) [reply docs/contributing/features.md]: Added the link. 

Right now, we're using the "enhancement" label to flag features people would like considered for addition to the Dockerfiles. I always assumed users could not add labels to issues unless they have read/write access to the repo. Is that incorrect?

File: docs/contributing/packages.md (hunk 7 above)
Comment (parente) [reply docs/contributing/packages.md]: I can extend the description of where the Dockerfiles reside in the repo.

File: docs/using/recipes.md (hunk 8 above)
Comment (parente) [reply docs/using/recipes.md]: Probably not. It didn't on the wiki either, surprisingly: https://github.com/jupyter/docker-stacks/wiki/Docker-recipes

File: docs/using/recipes.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+Ref: [https://github.com/jupyter/docker-stacks/issues/43](https://github.com/jupyter/docker-stacks/issues/43)
+
+## Let's Encrypt a Notebook server
+
+See the README for the simple automation here [https://github.com/jupyter/docker-stacks/tree/master/examples/make-deploy](https://github.com/jupyter/docker-stacks/tree/master/examples/make-deploy) which includes steps for requesting and renewing a Let's Encrypt certificate.
+
+Ref: [https://github.com/jupyter/docker-stacks/issues/78](https://github.com/jupyter/docker-stacks/issues/78)
+
+## Add Incubating Dashboard, Declarative Widget, Content Management Extensions
```
[the remark this answers is not in the document]
Comment (parente) [reply docs/using/recipes.md]: Removed the recipe. It's out of date and not particularly useful anymore.

File: docs/contributing/features.md (hunk 1 above)
Comment (parente) [reply docs/contributing/features.md]: Put bluntly, every feature requested for inclusion in the images means additional work by a very small number of people to maintain that feature over time. We have to be careful that the features we're agreeing to add are useful enough to justify their inclusion and long-term maintenance here. If we allow everything into these images, they'll become unwieldy to both maintain and use. Personally, I think we're past that point and need to take a more critical eye to what we add directly to this repository while continuing to let the ecosystem grow in other ways (e.g., through lower-maintenance recipes in the documentation, through third-party projects defining additional images).

I think my above commentary is *too* blunt for documentation, however. I don't want to scare off would-be contributors, just make them aware that we'd like to consider an array of options for capturing their features, not just dumping everything into uber-Docker images.

File: docs/using/recipes.md (hunk 8 above)
Comment (jzf2101) [reply docs/using/recipes.md]: Do we want to make this a link instead then?  Are you planing on keeping this in MD or converting to RST for the website?

File: docs/contributing/features.md (hunk 1 above)
Comment (jzf2101) [reply docs/contributing/features.md]: This is my suggested rephrasing:

>We review pull requests of new features (e.g., new packages, new scripts, new flags) to balance the value of the images to the Jupyter community with the cost of maintaining the images over time.  Roughly speaking, we evaluate these pull requests based on the following criteria:

File: docs/contributing/features.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+Please follow the process below to suggest a new feature for inclusion in one of the core stacks:
+
+1. [Open a GitHub issue](https://github.com/jupyter/docker-stacks/issues) describing the feature you'd like to contribute.
+2. Discuss with the maintainers whether the addition makes sense in [one of the core stacks](../using/selecting.html#Core-Stacks), as a [recipe in the documentation](recipes.html), as a [community stack](stacks.html), or as something else entirely.
+
+If there's agreement that the feature belongs in one or more of the core stacks:
+
+1. Implement the feature in a local clone of the `jupyter/docker-stacks` project.
+2. Try to build the image locally (`make image/somestack-notebook`) before submitting a pull request.
```
Comment (jzf2101) [inline docs/contributing/features.md]: Not sure about this but my guess is that this sentence should end with a colon for the bullets following?

File: docs/contributing/packages.md [hunk 14]
```diff
@@ -1,5 +1,14 @@
...
+# Package Updates
 
-## Package Updates
+We welcome all contributions that update packages already included in the project Dockerfiles. Please follow the process below to update a package version:
 
-## New Packages
\ No newline at end of file
+1. Locate the Dockerfile containing the library you wish to update (e.g., [base-notebook/Dockerfile](https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile), [scipy-notebook/Dockerfile](https://github.com/jupyter/docker-stacks/blob/master/scipy-notebook/Dockerfile))
+2. Adjust the version number for the package.
+    * We prefer to pin the major and minor version number of packages so as to minimize rebuild side-effects when users submit pull requests (PRs). For example, you'll find the Jupyter Notebook package, `notebook`, installed using conda with  `notebook=5.4.*`.
```
Comment (jzf2101) [inline docs/contributing/packages.md]: Given the fact that this is only 1 bullet could we merge this with line 6?

**Comment by jzf2101** [comment]
I'm going to make some additional suggestions to language and organization- if you'd prefer me to push to the pr let me know

File: docs/using/recipes.md [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+
+Here is a [quick example NGINX configuration](https://gist.github.com/cboettig/8643341bd3c93b62b5c2) to get started.  You'll need a server, a `.crt` and `.key` file for your server, and `docker` & `docker-compose` installed.  Then just download the files at that gist and run `docker-compose up -d` to test it out.  Customize the `nginx.conf` file to set the desired paths and add other services.
+
+## Using spark-packages.org
+
+If you'd like to use packages from [spark-packages.org](https://spark-packages.org/), see [https://gist.github.com/parente/c95fdaba5a9a066efaab](https://gist.github.com/parente/c95fdaba5a9a066efaab) for an example of how to specify the package identifier in the environment before creating a SparkContext.
+
+Ref: [https://github.com/jupyter/docker-stacks/issues/43](https://github.com/jupyter/docker-stacks/issues/43)
+
+## Let's Encrypt a Notebook server
```
Comment (jzf2101) [inline docs/using/recipes.md]: This could be moved to a security section perhaps?

**Comment by parente** [comment]
@jzf2101 Feel free to keep commenting or pushing to the PR as you see fit. I plan on getting back to making changes, but probably not til the weekend at the earliest.

**Comment by jzf2101** [comment]
Just did a reorganization and some additional language tweaking.  Let me know what you think @parente

**Comment by parente** [comment]
I had a few minutes to skim the changes you made @jzf2101. They look great so far! Thank you for taking the time to make them.

I'll have time to read through them more thoroughly this evening and leave feedback if I have any.

File: docs/contributing/features.md [hunk 16]
```diff
@@ -1,24 +1,33 @@
 # New Features
 
-We review contributions of new features (e.g., new packages, new scripts, new flags) to the core stacks in an attempt to balance the value of the images to the Jupyter community with the cost of maintaining the images over time. Roughly speaking, we think about the following when considering additions:
+Thank you for contributing to docker-stacks! We review pull requests of new features (e.g., new packages, new scripts, new flags) to balance the value of the images to the Jupyter community with the cost of maintaining the images over time.
+
+## Suggesting a New Feature
```
Comment (parente) [inline docs/contributing/features.md]: I like the suggest-evaluate-submit workflow. Your earlier comment about drive-by PR still stands. In those cases, I think we can treat the PR as the suggestion, albeit with initial implementation, point to the contributor to this  documentation, and then continue with the evaluation.

**Comment by parente** [comment]
@jzf2101 I think the updates look good. I'm in favor of merging what we have here and starting new PRs to work on the final two contribution sections (tests and community stacks) and point all of the READMEs to the docs site. Are you OK with that or still making changes?

We've had a handful of PRs merge with new features that were never documented because of the interim dual documentation in READMEs and on RTD. The sooner we get out of that state, the better I think, even if the RTD site still needs work.

**Comment by jzf2101** [comment]
Works for me
